src/librustc_hir/hir.rs

```diff
@@ -1330,6 +1330,18 @@ pub struct Expr<'hir> {
     pub span: Span,
 }
 
+/// `Expr` pattern alias to facilitate pattern-matching.
+///
+/// Usage examples:
+/// * `hir::Expr!(Unary(_, hir::Expr!(Lit(_))))`
+/// * `hir::Expr! { Loop(..), hir_id: loop_id }`
+pub macro Expr($variant:ident $($rest:tt)*) {
```

Comment:
why not use `pat` fragment?

Reply:
Because then people would have to write `hir::ExprKind::Foo` instead of `Foo`, inside the macro arguments.
And also this allows some trickery like matching on the `hir_id` (although maybe that should be done with `@` instead of this).